### PR TITLE
Svg parsing refactoring

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**
This pull request refactors/prettyfies much of the SVG parsing/generation code.

Refactoring of the parsing is not yet complete. I plan to do this incrementally throughout the week. If you have any concerns or comments, please let me know.

### CODE REVIEW

File: hs/SvgParsing/ParserUtil.hs [hunk 1]
```diff
@@ -74,27 +74,27 @@ getChildren = path [children]
...
 -- | Parses a path's `d` attribute.
 parsePathD :: String -> [(Float, Float)]
-parsePathD d =
-    if head d == 'm'
-    then foldCoordsRel $ filter (\x -> length x > 1) $ map (splitOn ",") $ splitOn " " d
-    else processAbsCoords $ filter (\x -> length x > 1) $ map (splitOn ",") $ splitOn " " d
+parsePathD d
+    | head d == 'm' = foldCoordsRel $ filter lengthMoreThanOne $ map (splitOn ",") $ splitOn " " d
+    | otherwise =  processAbsCoords $ filter lengthMoreThanOne $ map (splitOn ",") $ splitOn " " d
+    where lengthMoreThanOne = (\x -> length x > 1)
```
Comment (david-yz-liu) [inline hs/SvgParsing/ParserUtil.hs]: You don't need the parentheses here. Also, put the duplicated argument in the where.

File: hs/SvgParsing/SVGBuilder.hs [hunk 2]
```diff
@@ -16,32 +16,30 @@ import SvgParsing.ParserUtil
...
-             (pathsIsRegion entity)
-             sourceNode
-             targetNode
+    | otherwise =
+        let xStart = fst $ head coords
+            yStart = snd $ head coords
+            xEnd = fst $ last coords
+            yEnd = snd $ last coords
+            sourceNode = getIntersectingShape xStart yStart (rects ++ ellipses)
+            targetNode = getIntersectingShape xEnd yEnd (rects ++ ellipses) in
```
Comment (david-yz-liu) [inline hs/SvgParsing/SVGBuilder.hs]: Put the `in` on the next line

File: hs/SvgParsing/SVGBuilder.hs [hunk 3]
```diff
@@ -129,8 +123,10 @@ buildPathString d = unwords $ map (joinPathTuple . convertRationalTupToString) d
 
 -- | Joins two String values in a tuple with a comma.
 joinPathTuple :: (String, String) -> String
-joinPathTuple tup = fst tup ++ "," ++ snd tup
+joinPathTuple (a, b) = a ++ "," ++ b
 
 -- | Converts a tuple of Rationals to a tuple of String.
+-- NOTE: `convertRationalTupToString (a, b) = (show $ fromRational a, show $ fromRational a)`
```
Comment (david-yz-liu) [inline hs/SvgParsing/SVGBuilder.hs]: Can you clarify why this won't work?

File: hs/SvgParsing/SVGGenerator.hs [hunk 4]
```diff
@@ -188,4 +190,11 @@ convertEllipseToSVG ellipse = S.g ! A.id_ (stringValue (shapeId ellipse))
                                                 ! A.rx (stringValue $ show $ fromRational $ shapeWidth ellipse / 2)
                                                 ! A.ry (stringValue $ show $ fromRational $ shapeHeight ellipse / 2)
                                                 ! A.style "stroke:#000000;fill:none;"
-                                      concatSVG $ map (convertTextToSVG False True False) (shapeText ellipse)
\ No newline at end of file
+                                      concatSVG $ map (convertTextToSVG False True False) (shapeText ellipse)
+
+getTextStyle :: Bool -> Bool -> Bool -> String
+getTextStyle isHybrid isBool isRegion
```
Comment (david-yz-liu) [inline hs/SvgParsing/SVGGenerator.hs]: I'd prefer to see pattern-matching here

File: hs/SvgParsing/SVGBuilder.hs (hunk 2 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/SVGBuilder.hs]: Will do. I'll check if this happens anywhere else, as well.

File: hs/SvgParsing/SVGBuilder.hs (hunk 3 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/SVGBuilder.hs]: `a`.

File: hs/SvgParsing/ParserUtil.hs (hunk 1 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/ParserUtil.hs]: Will do.

File: hs/SvgParsing/SVGGenerator.hs (hunk 4 above)
Comment (Ian-Stewart-Binks) [inline hs/SvgParsing/SVGGenerator.hs]: As in 
```
getTextStyle :: Bool -> Bool -> Bool -> String
getTextStyle True True False = hybridTextStyle
getTextStyle False True False = ellipseTextStyle
getTextStyle False False True = regionTextStyle
getTextStyle _ _ _ = ""
```
?

Is there a better way we can approach separating texts that are in bools vs rects vs regions?

**Comment by david-yz-liu** [comment]
Passing in the three booleans is pretty silly. A simpler approach (though maybe not the best one) is to use an Enum of four values.

**Pull request merged**